### PR TITLE
Low prio: check if the two operands are of the same type before interpreting a missing prop as a remove op

### LINKED ISSUE #31: Wrong patch when substituting an object for an array?
```
> jsonpatch.compare({a: {}}, {a: [1]});
[{op: "add", path: "/a/0", value: 1}]
```

Shouldn't be the empty object replaced with the array? the value is no longer an object.

```
[{op: "replace", path: "/a", value: [1]}]
```

### PR DESCRIPTION
**Pull request by alshakero**
Fixes #31

### CODE REVIEW

File: test/spec/duplexSpec.js [hunk 1]
```diff
@@ -1426,8 +1426,47 @@ describe('duplex', function() {
...
+      expect(patches).toEqual([
+        {
+          op: 'replace',
+          path: '',
+          value: obj
+        }
+      ]);
+
+    });
+    it('Replacing a deep array with an object should be handled well', function() {
```
Comment (tomalec) [inline test/spec/duplexSpec.js]: It's rather an array nested in an object, than deep array, as it's only shallow - one level array of primitives.

File: test/spec/duplexSpec.js [hunk 2]
```diff
@@ -1426,8 +1426,47 @@ describe('duplex', function() {
...
+      expect(patches).toEqual([
+        {
+          op: 'replace',
+          path: '/arr',
+          value: obj
+        }
+      ]);
+            
+    });
+    it('Replacing a two-level deep array with an object should be handled well', function() {
```
Comment (tomalec) [inline test/spec/duplexSpec.js]: as in https://github.com/Starcounter-Jack/JSON-Patch/pull/205/files#r192806464

File: src/duplex.ts [hunk 3]
```diff
@@ -217,9 +219,12 @@ function _generate(mirror, obj, patches, path) {
         }
       }
     }
-    else {
+    else if(Array.isArray(mirror) === Array.isArray(obj)) {
```
Comment (tomalec) [inline src/duplex.ts]: Could you add some code comments about what's happening here? Why for two arrays or two non-arrays it's a `delete`, but for array and non-array, it's a `replace`?

Comment (alshakero) [reply src/duplex.ts]: I'll add something of the effect of: Because the `replace` is for the whole obj, not for `obj[key]`.

**Comment by robertdumitrescu** [comment]
When will be this merged? I created a test in mocha and the array replacement is not the only problem. 
`
it('should getDiff and by applying the diff changes to get the modified object', async () => {

            let firstNestedObject = [
                {
                    title: 'Products',
                    title2: 'Products2',
                    children: [
                        {
                            title: 'Electronics',
                            children: [
                                {
                                    title: 'Audio-Video',
                                    children: [
                                        {
                                            title: 'Game consoles',
                                            children: [
                                                {
                                                    title: 'Xbox 360',
                                                    date: new Date(2018, 11, 24, 10, 33, 30)
                                                },
                                                {
                                                    title: 'PlayStation 4',
                                                },
                                                {
                                                    title: 'Sega',
                                                }
                                            ]
                                        }
                                    ]
                                },
                                {
                                    title: 'Mobile devices'
                                }
                            ]
                        },
                        {
                            title: 'Cars'
                        }
                    ]
                },
                {
                    title: 'News',
                    randomDate: new Date(2018, 11, 24, 10, 33, 30),
                    children: [
                        {
                            title: 'Technology',
                            children: [
                                {
                                    title: 'CES 2018'
                                },
                                {
                                    title: 'Why PHP sucks'
                                }
                            ]
                        },
                        {
                            title: 'Environment',
                            subObj: {
                                prop1: 'randomProp'
                            }
                        }
                    ]
                }
            ];


            let secondModifiedNestedObject = [
                {
                    title: 'Products Modified - =====>HEEEREE',
                    title2: false,
                    children: [
                        {
                            title: 'Electronics',
                            children: [
                                {
                                    title: 'Audio-Video',
                                    extraProp: 1,
                                    extraProp00: 1413431431431413,
                                    extraProp2: [
                                        'bla',
                                        {
                                            extraSubSneakyProp: 'Super Sneaky Prop =====> HEEEERE'
                                        }
                                    ],
                                    children: [
                                        {
                                            title: 'Game consoles',
                                            children: [
                                                {
                                                    title: 'Xbox 360',
                                                    date: null
                                                },
                                                {
                                                    title22: 'Sega',
                                                }
                                            ]
                                        }
                                    ]
                                },
                                {
                                    title: 'Mobile devices'
                                }
                            ]
                        },
                        {
                            title: true
                        }
                    ]
                },
                {
                    title: 'News',
                    randomDate: new Date(2018, 11, 9, 10, 36, 30),
                    children: [
                        {
                            title: null,
                            children: [
                                {
                                    title: 'CES 2018'
                                },
                                {
                                    title: 'Why PHP sucks'
                                },
                                {
                                    title: 'Some new article,',
                                    customSUbArray: [
                                        {
                                            subObject: 'nananana',
                                            _weirdProp: 'nanana2'
                                        }
                                    ]
                                }
                            ]
                        },
                        {
                            title: 'Environment',
                            subObj: [
                                1, 3, 5
                            ]
                        }
                    ]
                }
            ];

            let differences = GenericObjectHelper.getRFC6902Diff(firstNestedObject, secondModifiedNestedObject);
            let actualModifiedObject = GenericObjectHelper.applyRFC6902Diff(firstNestedObject, differences);

            expect(actualModifiedObject).to.deep.equal(secondModifiedNestedObject);
        });
`

And it fails on both getting the diff out of a Date object and also when an object is replace with an array. 
I am currently doing a spike for this package since we need to develop something similar to git versioning but for Elastic Search documents. Any update?

**Comment by robertdumitrescu** [comment]
Ah... I saw your comments in another issue about Date objects. Fair enough. Is not the focus of this plugin. Apologies. But the array/object replace is a must have though.

**Pull request merged**